### DISCUSSION TITLE
StrongSwan key exchange

### DISCUSSION THREAD

**Discussion by musaleh17** [strongswan/strongswan#1519] [Q&A]
Hi,

I want to use strongswang for only keys exchange between two points which creates the IPSEC SA without running ipsec lib.
As I saw that libcharon uses the IPSEC kernel stack by default. 

is it already supported in strongswan plugins? without running IPSEC flows and kernel IPSEC stack? 
or should  I add/implement my own plugin?

**Comment by tobiasbrunner** [comment]
> which creates the IPSEC SA without running ipsec lib

What does that mean exactly? What "ipsec lib"?

> As I saw that libcharon uses the IPSEC kernel stack by default.

It depends on the plugins. But yes, the default is the _kernel-netlink_ plugin, which installs SAs in the Linux kernel via XFRM.

> is it already supported in strongswan plugins? without running IPSEC flows and kernel IPSEC stack?

What do you mean? Supported how?

> or should I add/implement my own plugin?

Maybe, depends on what you want to do. Implementing the `kernel_ipsec_t` interface in a custom plugin could definitely be an option.

**Reply by musaleh17** [reply]
>>Maybe, depends on what you want to do. Implementing the kernel_ipsec_t interface in a custom plugin could definitely be an option.
> In this case, will I still need the kernel-netlink in order to load my plugin? 

>Mainly I won't install SAs to the Linux kernel, is it possible? 
> Does strongswan have a plugin for something similar?

**Reply by musaleh17** [reply]
when I am trying to disable kernel-netlink I have this error 

>>>
Selecting previously unselected package strongswan.
(Reading database ... 81539 files and directories currently installed.)
Preparing to unpack strongswan_5.9.6-2bf_arm64.deb ...
Unpacking strongswan (5.9.6-2bf) ...
Setting up strongswan (5.9.6-2bf) ...
Created symlink /etc/systemd/system/multi-user.target.wants/strongswan-starter.service → /lib/systemd/system/strongswan-starter.service.
Created symlink /etc/systemd/system/strongswan-swanctl.service → /lib/systemd/system/strongswan.service.
Created symlink /etc/systemd/system/multi-user.target.wants/strongswan.service → /lib/systemd/system/strongswan.service.
**Job for strongswan.service failed because the control process exited with error code.**
See "systemctl status strongswan.service" and "journalctl -xe" for details.
Processing triggers for libc-bin (2.31-0ubuntu9.9) ...
Processing triggers for man-db (2.9.1-1) ...

>>>
systemctl status strongswan.service
● strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl
     Loaded: loaded (/lib/systemd/system/strongswan.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2023-01-24 13:57:03 UTC; 1min 31s ago
    Process: 2953011 ExecStart=/usr/sbin/charon-systemd (code=exited, status=66)
   Main PID: 2953011 (code=exited, status=66)
     Status: "charon initialization failed"

Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading ca certificates from '/etc/ipsec.d/cacerts'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading aa certificates from '/etc/ipsec.d/aacerts'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading attribute certificates from '/etc/ipsec.d/acerts'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading crls from '/etc/ipsec.d/crls'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: loading secrets from '/etc/ipsec.secrets'
Jan 24 13:57:03 dev-doca-45-044-bf charon-systemd[2953011]: failed to load 1 critical plugin feature
Jan 24 13:57:03 dev-doca-45-044-bf systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
Jan 24 13:57:03 dev-doca-45-044-bf systemd[1]: strongswan.service: Failed with result 'exit-code'.
Jan 24 13:57:03 dev-doca-45-044-bf systemd[1]: Failed to start strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.

**Reply by tobiasbrunner** [reply]
> In this case, will I still need the kernel-netlink in order to load my plugin?

Not to load your plugin, but you might need the `kernel_net_t` implementation of the _kernel-netlink_ plugin to handle networking. Just make sure to load your own plugin that registers a `kernel_ipsec_t` implementation before the _kernel-netlink_ plugin.

> Does strongswan have a plugin for something similar?

Just search for `kernel_ipsec_t` implementations in the code. The one by the _load-tester_ plugin is kind of a no-op implementation (as is the one used for the _libcharon_ unit tests).

**Reply by musaleh17** [reply]
>>Just make sure to load your own plugin that registers a kernel_ipsec_t implementation before the kernel-netlink plugin

> How should I do that?

**Reply by tobiasbrunner** [reply]
[Read the docs](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html).

**Reply by musaleh17** [reply]
I did as follow:
1) enable charon.load_modular = yes
2) set my loading plugin with a priority greater than 1; charon.plugins.doca.load := 2
3) I build the deb files for strongswan and install them on my device 
when I ran  systemctl status strongswan-starter.service
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[CFG] loading crls from '/etc/ipsec.d/crls'
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[CFG] loading secrets from '/etc/ipsec.secrets'
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[LIB] loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 >
Jan 24 15:31:19 dev-doca-45-044-bf charon[3085553]: 00[JOB] spawning 16 worker threads
Jan 24 15:31:19 dev-doca-45-044-bf ipsec[3085537]: charon (3085553) started after 40 ms
Jan 24 15:31:19 dev-doca-45-044-bf ipsec_starter[3085537]: charon (3085553) started after 40 ms
Jan 24 15:31:21 dev-doca-45-044-bf systemd[1]: Started strongSwan IPsec IKEv1/IKEv2 daemon using ipsec.conf.


I don't see that my pluugin was loaded! 
Do you know what might be the issue?

**Reply by tobiasbrunner** [reply]
There are so many things that could be wrong (config incorrect, plugin incorrectly built, not installed, installed in the wrong location, doing the wrong thing etc.). Try reading the full log (maybe configure [higher log levels](https://docs.strongswan.org/docs/5.9/config/logging.html)).

Also, you write "deb files". Combining custom plugins with binary distribution packages is very tricky. This only works if you use the exact same source, build environment and configure options that were used to build those binaries (strongSwan does not provide any stable APIs).

**Reply by musaleh17** [reply]
When running the debuild command I see my plugin was been compiled and installed 
Making install in plugins/updown
make[7]: Entering directory '/root/strongswan/src/libcharon/plugins/updown'
make[8]: Entering directory '/root/strongswan/src/libcharon/plugins/updown'
make[8]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
 /bin/bash ../../../../libtool   --mode=install /usr/bin/install -c   libstrongswan-updown.la '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
libtool: install: /usr/bin/install -c .libs/libstrongswan-updown.so /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-updown.so
libtool: install: /usr/bin/install -c .libs/libstrongswan-updown.lai /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-updown.la
libtool: warning: remember to run 'libtool --finish /usr/lib/ipsec/plugins'
make[8]: Leaving directory '/root/strongswan/src/libcharon/plugins/updown'
make[7]: Leaving directory '/root/strongswan/src/libcharon/plugins/updown'
Making install in plugins/kernel_netlink
make[7]: Entering directory '/root/strongswan/src/libcharon/plugins/kernel_netlink'
make[8]: Entering directory '/root/strongswan/src/libcharon/plugins/kernel_netlink'
make[8]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
 /bin/bash ../../../../libtool   --mode=install /usr/bin/install -c   libstrongswan-kernel-netlink.la '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
libtool: install: /usr/bin/install -c .libs/libstrongswan-kernel-netlink.so /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-kernel-netlink.so
libtool: install: /usr/bin/install -c .libs/libstrongswan-kernel-netlink.lai /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-kernel-netlink.la
libtool: warning: remember to run 'libtool --finish /usr/lib/ipsec/plugins'
make[8]: Leaving directory '/root/strongswan/src/libcharon/plugins/kernel_netlink'
make[7]: Leaving directory '/root/strongswan/src/libcharon/plugins/kernel_netlink'
Making install in plugins/doca
make[7]: Entering directory '/root/strongswan/src/libcharon/plugins/doca'
make[8]: Entering directory '/root/strongswan/src/libcharon/plugins/doca'
make[8]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
 /bin/bash ../../../../libtool   --mode=install /usr/bin/install -c   libstrongswan-doca.la '/root/strongswan/debian/tmp/usr/lib/ipsec/plugins'
libtool: install: /usr/bin/install -c .libs/libstrongswan-doca.so /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-doca.so
libtool: install: /usr/bin/install -c .libs/libstrongswan-doca.lai /root/strongswan/debian/tmp/usr/lib/ipsec/plugins/libstrongswan-doca.la
libtool: warning: remember to run 'libtool --finish /usr/lib/ipsec/plugins'
**make[8]: Leaving directory '/root/strongswan/src/libcharon/plugins/doca'
make[7]: Leaving directory '/root/strongswan/src/libcharon/plugins/doca'**


also the .so .la files exist under /usr/lib/ipsec/plugins/
so, I am not sure what is missing ...

**Reply by musaleh17** [reply]
>>>Also, you write "deb files". Combining custom plugins with binary distribution packages is very tricky. This only works if you use the exact same source, build environment and configure options that were used to build those binaries (strongSwan does not provide any stable APIs).

> I am building and installing the package on the same device, so no worries from this point.

> also the weird thing is when I run ./configure locally without running debuild I see my plugin is enabled
libcharon:     attr doca kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
